src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java

```diff
@@ -650,6 +650,19 @@ void cleanup(Control[] reqCtls, boolean notifyParent) {
                     } catch (IOException ie) {
                         if (debug)
                             System.err.println("Connection: problem closing socket: " + ie);
+                        //bug 8313657, socket not closed util GC running
+                        try {
+                            sock.close();
+                            unpauseReader();
+                            if (debug) {
+                                if (sock.isClosed()) {
+                                    System.err.println("Connection::cleanup - socket closed.");
+                                }
+                            }
+                        } catch (IOException ioe) {
+                            if (debug)
+                                System.err.println("Connection::cleanup problem: " + ioe);
```

Comment:
Maybe change the message here to highlight the fact that it's the 2nd attemp to clean-up the connection